nucypher/blockchain/eth/actors.py

```diff
@@ -440,27 +407,74 @@ def _is_phase_1_action_required(self, ritual_id: int) -> bool:
         timestamp: int,
-    ) -> Optional[HexBytes]:
-        """Perform round 1 of the DKG protocol for a given ritual ID on this node."""
+    ) -> Optional[AsyncTx]:
+        """
+        Perform phase 1 of the DKG protocol for a given ritual ID on this node.
+
+        This method is idempotent and will not submit a transcript if one has
+        already been submitted. It is dispatched by the EventActuator when it
+        receives a StartRitual event from the blockchain.  Since the EventActuator
+        scans overlapping blocks, it is possible that this method will be called
+        multiple times for the same ritual.  This method will check the state of
+        the ritual and participant on the blockchain before submitting a transcript.
+
+        If a there is a tracked AsyncTx for the given ritual and round
+        combination, this method will return the tracked transaction.  If there is
+        no tracked transaction, this method will submit a transcript and return the
+        resulting FutureTx.
+
+        Returning None indicates that no action was required or taken.
+
+        Errors raised by this method are not explicitly caught and are expected
+        to be handled by the EventActuator.
+        """
         if self.checksum_address not in participants:
-            # should never get here
-            self.log.error(
-                f"Not part of ritual {ritual_id}; no need to submit transcripts"
-            )
-            raise RuntimeError(
-                f"Not participating in ritual {ritual_id}; should not have been notified"
+            # ERROR: This is an abnormal state since this method
+            # is designed to be invoked only when this node
+            # is an on-chain participant in the Coordinator.StartRitual event.
+            #
+            # This is a *nearly* a critical error.  It's possible that the
+            # log it as an error and return None to avoid crashing upstack
+            # async tasks and drawing unnecessary amounts of attention to the issue.
+            message = (
+                f"{self.checksum_address}|{self.wallet_address} "
+                f"is not a member of ritual {ritual_id}"
             )
+            self.log.error(message)
+            return
 
         # check phase 1 contract state
         if not self._is_phase_1_action_required(ritual_id=ritual_id):
+            self.log.debug(
+                "No action required for phase 1 of DKG protocol for some reason or another."
+            )
             return
 
+        # check if there is  already pending tx for this ritual + round combination
+        async_tx = self.ritual_tracker.active_rituals.get(
+            self._phase_id(ritual_id, PHASE1)
+        )
+        if async_tx:
+            self.log.info(
+                f"Active ritual in progress: {self.transacting_power.account} has submitted tx"
+                f"for ritual #{ritual_id}, phase #{PHASE1} (final: {async_tx.final})"
```

Comment:
what is async_tx.final? I sthat the final tx hash for completion, or something else?

Reply:
Excellent question - it's a boolean. This is part of the `atxm` API that allows callers to block until an async transaction is final.

Reply:
One reason for the question is, is there an edge case we need to be worried about...maybe...? 

Presumably final being True means that the transaction has completed in some way (successful or not, but completed). If we get here then the contract indicates that action is needed by the node, but if the async tx is final yet the contract still claims that the node needs to act, what do we do then? I may be misunderstanding the asynctx library...

Reply:
Another excellent question - these are the types of questions I am trying to answer in atxm.  final will only be true here if the transaction has actually been included in a block, however this object may have a faulted state instead of a final state.  These faults are strategy-specific and can be handled here in nucypher at this layer.  For example, what shall we do with a reverted transaction?  I am not trying to resolve that question in this PR, rather, establish a platform where can can build a proper response.

There are indeed edge cases!